### PR TITLE
Feature/sane add plugin v3

### PR DESCRIPTION
**Pull request by czpython**
* Removed ``admin/cms/plugin/close_modal.html`` because it's logic is already implemented in ``admin/cms/page/plugin/confirm_form.html``.
* Refactored the add request validation logic on the plugin base class
* Fixed backwards compatibility with projects that have these ghost plugins already created.
* Cleaned up code that used to handle cancellation of editing a plugin.
* Switched ckeditor test to pull from specific sane-add-plugin branch /cc @mkoistinen

### CODE REVIEW

**Comment by czpython** [comment]
@vxsx I see you've merged develop.. this makes it very complicated to review, can you rebase the sane-add-plugin branch with develop?

**Comment by vxsx** [comment]
@czpython sorry, fixed

**Comment by czpython** [comment]
@ojii hey, so this is and https://github.com/divio/djangocms-text-ckeditor/pull/312 are the last prs needed to merge sane-add-plugin into CMS develop. Do you mind reviewing both?

Would love feedback from @divio/django-cms-core as well.

File: cms/static/cms/js/modules/cms.base.js [hunk 1]
```diff
@@ -523,11 +523,14 @@ var CMS = {
         /**
          * Dispatches an event
          * @method dispatchEvent
-         * @param {String} eventName  string containing space separated event names
+         * @param {String} eventName event name
          * @param {Object} payload whatever payload required for the consumer
+         * @returns {$.Event} event that was just triggered;
```
Comment (FinalAngel) [inline cms/static/cms/js/modules/cms.base.js]: remove ";"

File: cms/templates/admin/cms/page/close_frame.html [hunk 2]
```diff
@@ -11,16 +11,17 @@
...
     // the dataBridge is used to access plugin information from different ressources
+    // Do NOT move this!!!
     Window.CMS.API.Helpers.dataBridge = {
         'plugin_id': '{{ plugin_id|unlocalize }}',
         'plugin_type': '{{ type }}',
         'plugin_name': '{{ name }}',
         'plugin_icon': '{{ icon }}',
         'plugin_desc': '{{ alt }}'
     };
+// CMS.$ will be passed for $
```
Comment (FinalAngel) [inline cms/templates/admin/cms/page/close_frame.html]: remove comment

**Comment by ojii** [comment]
I don't have the time to in-depth review this, but it looks like all the sane bits are there and doing what they're supposed to do. Fantastic work everyone!

File: cms/templates/admin/cms/page/close_frame.html [hunk 3]
```diff
@@ -10,22 +10,23 @@
...
-    // save current plugin
-    Window.CMS.API.Helpers.reloadBrowser(undefined, {{ window_close_timeout|default:500 }});
-});
-// we have a special case here cause the CMS namespace
-// can be either inside the current window or the parent
-})(window.parent || window);
+    // we have a special case here cause the CMS namespace
+    // can be either inside the current window or the parent
+    (function(Window) {
+        // the dataBridge is used to access plugin information from different ressources
```
Comment (mkoistinen) [inline cms/templates/admin/cms/page/close_frame.html]: "resources"

**Comment by mkoistinen** [comment]
Looks really good! Great job. OK to merge.

**Comment by yakky** [comment]
LGTM

**Comment by ojii** [comment]
> czpython merged commit 2aa18e4 into divio:sane-add-plugin 25 minutes ago

what a time to be alive

**Pull request merged**